src/lib/SaferNames/Builder.hs

```diff
@@ -0,0 +1,296 @@
+    return $ DistinctAbs Empty getDistinctEvidence
+
+-- === BuilderT ===
+
+newtype BuilderT (m::MonadKind1) (n::S) (l::S) (a:: *) =
+  BuilderT { runBuilderT' :: Inplace (BuilderNameGenT Decl m) n l a }
+  deriving (Functor, Applicative, Monad)
+
+runBuilderT :: (BindingsReader m, BindingsExtender m, InjectableE e)
+            => (forall l. (Distinct l, Ext n l) => BuilderT m n l (e l))
+            -> m n (DistinctAbs (Nest Decl) e n)
+runBuilderT cont = runBuilderNameGenT $ runInplace $ runBuilderT' cont
+
+instance (BindingsReader m, BindingsExtender m)
+         => Builder (BuilderT m n) where
+  emitDecl hint ann expr = BuilderT $
+    liftInplace $ BuilderNameGenT do
+      expr' <- injectM expr
+      ty <- getType expr'
+      let binderInfo = LetBound ann expr'
+      withFreshBinder hint ty binderInfo \b -> do
+        return $ DistinctAbs (Nest (Let ann b expr') Empty) (binderName b)
+
+apBuilder :: (InjectableE e, InjectableE e')
+          => (forall l'. e l' -> m l' (e' l'))
+          -> e l
+          -> BuilderT m n l (e' l)
+apBuilder cont x = liftBuilder x cont
+
+liftBuilder :: (InjectableE e, InjectableE e')
+            => e l
+            -> (forall l'. e l' -> m l' (e' l'))
+            -> BuilderT m n l (e' l)
+liftBuilder _ _ = undefined
+
+-- === effects monad ===
+
+class Monad1 m => EffectsReader m where
+  getAllowedEffects :: m n (EffectRow n)
+
+newtype EffectsReaderT (m::MonadKind1) (n::S) (a:: *) =
+  EffectsReaderT { runEffectsReaderT' :: ReaderT (EffectRow n) (m n) a }
+  deriving (Functor, Applicative, Monad)
+
+instance ScopeReader m => (ScopeReader (EffectsReaderT m))
+
+instance BindingsReader m => (BindingsReader (EffectsReaderT m))
+
+instance Monad1 m => EffectsReader (EffectsReaderT m)
+
+runEffectsReaderT :: EffectRow n -> EffectsReaderT m n a -> m n a
+runEffectsReaderT effs cont = runReaderT (runEffectsReaderT' cont) effs
+
+-- === lambda-like things ===
+
+buildBlock :: (BindingsReader m, BindingsExtender m, MonadFail1 m)
+            => (forall l. (Distinct l, Ext n l) => BuilderT m n l (Atom l))
+            -> m n (Block n)
+buildBlock cont = do
+  DistinctAbs decls (PairE result ty) <- runBuilderT do
+    result <- cont
+    ty <- getType `apBuilder` result
+    return $ PairE result ty
+  ty' <- fromConstAbs $ Abs decls ty
+  return $ Block ty' decls $ Atom result
+
+atomAsBlock :: BindingsReader m => Atom n -> m n (Block n)
+atomAsBlock x = do
+  ty <- getType x
+  return $ Block ty Empty $ Atom x
+
+buildLamAux :: (EffectsReader m, BindingsReader m)
+            => Type n
+            -> Arrow
+            -> (forall l. Ext n l => m l (EffectRow l))
+            -> (forall l. Ext n l => m l (Block l, a))
+            -> m n (Atom n, a)
+buildLamAux ty arr fEff fBody = undefined
+
+buildLam :: (EffectsReader m, BindingsReader m)
+         => Arrow
+         -> Type n
+         -> (forall l. Ext n l => Atom l -> m l (Block l))
+         -> m n (Atom n)
+buildLam _ _ _ = undefined
```

Comment:
Yes, I'll try to finish most of the implementation before merging. But none of it will really get tested until we use it in inference, which will be the next PR.